### ISSUE TITLE
Test failure: CanVerifyPhoneNumber

### ISSUE THREAD

**Issue by ryanbrandenburg** [aspnet/Identity#1913]
This test [fails](http://aspnetci/viewLog.html?buildId=524549&buildTypeId=XPlat_Mac_MacUniverseCoreCLR) occasionally with the following error:
```
Assert.False() Failure
Expected: False
Actual:   True
   at Microsoft.AspNetCore.Identity.Test.UserManagerSpecificationTestBase`2.CanVerifyPhoneNumber() in /_/src/Specification.Tests/UserManagerSpecificationTests.cs:line 1682
--- End of stack trace from previous location where exception was thrown ---
```
Other tests within that build may have failed with a similar message, but they are not listed here. Check the link above for more info.

CC @Eilon,@muratg,@mkArtakMSFT

This issue was made automatically. If there is a problem contact @ryanbrandenburg.

**Comment by ryanbrandenburg** [comment]
Please use this workflow to address this flaky test issue, including checking applicable checkboxes and filling in the applicable "TODO" entries:

* Is this actually a flaky test?
  * No, this is a regular test failure, fix the test/product (TODO: Link to commit/PR)
  * Yes, proceed below...

* Is this test failure caused by product code flakiness? (Either this product, or another product this test depends on.)
  * [ ] File a bug against the product (TODO: Link to other bug)
  * Is it possible to change the test to avoid the flakiness?
    * Yes? Go to the "Change the test!" section.
    * No?
      * [ ] Disable the test (TODO: Link to PR/commit)
      * [ ] Wait for other bug to be resolved
      * [ ] Wait for us to get build that has the fix
      * [ ] Re-enable our test (TODO: Link to PR/commit)
      * [ ] Close this bug

* Is it that the test itself is flaky? This includes external transient problems (e.g. remote server problems, file system race condition, etc.)
  * Is there is a way to change our test to avoid this flakiness?
    * Yes? Change the test!
      * [ ] Change the test to avoid the flakiness, for example by using a different test strategy, or by adding retries w/ timeouts (TODO: Link to PR/commit)
      * [ ] Run the test 100 times locally as a sanity check.
      * [ ] Close this bug
    * No?
      * [ ] Delete the test because flaky tests are not useful (TODO: Link to PR/commit)

This comment was made automatically. If there is a problem contact @ryanbrandenburg.

**Comment by aspnet-hello** [comment]
EmailFactorFailsAfterSecurityStampChangeTest [failed with about the same error](http://aspnetci/viewLog.html?buildId=551277&buildTypeId=Releases_21Public_UbuntuUniverse) on 2.1.

This comment was made automatically. If there is a problem contact ryanbrandenburg.

**Comment by Eilon** [comment]
@HaoK - any idea on this?

**Comment by HaoK** [comment]
Its only failing on 2.1?

**Comment by Eilon** [comment]
Looks that way right now.

**Comment by HaoK** [comment]
Test looks unchanged for a while, so 2.0 -> 2.2 all should be the same, the last assert is failing which is swapping the tokens and making sure they don't work with the wrong phone number.  Relevant code included in case anything jumps out at anyone.  The test already checks to make sure the tokens are different, so its not a TOTP collision I don't think

```C#
        [Fact]
        public async Task CanVerifyPhoneNumber()
        {
            var manager = CreateManager();
            var user = CreateTestUser();
            IdentityResultAssert.IsSuccess(await manager.CreateAsync(user));
            const string num1 = "111-123-4567";
            const string num2 = "111-111-1111";
            var userId = await manager.GetUserIdAsync(user);
            var token1 = await manager.GenerateChangePhoneNumberTokenAsync(user, num1);

            var token2 = await manager.GenerateChangePhoneNumberTokenAsync(user, num2);
            Assert.NotEqual(token1, token2);
            Assert.True(await manager.VerifyChangePhoneNumberTokenAsync(user, token1, num1));
            Assert.True(await manager.VerifyChangePhoneNumberTokenAsync(user, token2, num2));
            Assert.False(await manager.VerifyChangePhoneNumberTokenAsync(user, token2, num1));
            Assert.False(await manager.VerifyChangePhoneNumberTokenAsync(user, token1, num2));

     // Phone number token is TOTP with phone number entrophy:

        public virtual async Task<string> GenerateAsync(string purpose, UserManager<TUser> manager, TUser user)
        {
            var token = await manager.CreateSecurityTokenAsync(user);
            var modifier = await GetUserModifierAsync(purpose, manager, user);
            return Rfc6238AuthenticationService.GenerateCode(token, modifier).ToString("D6", CultureInfo.InvariantCulture);
        }

        public override async Task<string> GetUserModifierAsync(string purpose, UserManager<TUser> manager, TUser user)
        {
            var phoneNumber = await manager.GetPhoneNumberAsync(user);
            return "PhoneNumber:" + purpose + ":" + phoneNumber;
        }

        public virtual async Task<bool> ValidateAsync(string purpose, string token, UserManager<TUser> manager, TUser user)
        {
            int code;
            if (!int.TryParse(token, out code))
            {
                return false;
            }
            var securityToken = await manager.CreateSecurityTokenAsync(user);
            var modifier = await GetUserModifierAsync(purpose, manager, user);
            return securityToken != null && Rfc6238AuthenticationService.ValidateCode(securityToken, code, modifier);
        }
```

**Comment by Eilon** [comment]
Yeah that's odd... Can we add some logging spew to it so that we can get more info about it if it fails again?

**Comment by Eilon** [comment]
(The point being, especially because this is odd, that we need to do *something*, just in case there's some deeper product bug here.)

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact ryanbrandenburg.

There were 2 failures [with about the same error](http://aspnetci/viewLog.html?buildId=585272&buildTypeId=XPlat_Mac_MacUniverseCoreCLR) on master at 1:50:55 PM:
- CanRequireConfirmedPhoneNumberForPasswordSignIn
- CanRequireConfirmedEmailForPasswordSignIn

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact ryanbrandenburg.

There were 2 failures [with about the same error](http://aspnetci/viewLog.html?buildId=585268&buildTypeId=XPlat_Windows_Win2008r2_Universe) on master at 1:50:55 PM:
- CanRequireConfirmedPhoneNumberForPasswordSignIn
- CanRequireConfirmedEmailForPasswordSignIn

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact ryanbrandenburg.

There were 2 failures [with about the same error](http://aspnetci/viewLog.html?buildId=585675&buildTypeId=XPlat_Windows_Win8_Universe) on master at 3:03:03 PM:
- CanRequireConfirmedEmailForPasswordSignIn
- CanRequireConfirmedPhoneNumberForPasswordSignIn
